pallets/marketplace/src/lib.rs

```diff
@@ -0,0 +1,512 @@
+                .map_err(|_| Error::<T>::InvalidPayment)?;
+
+            <StoredJobStatus<T>>::insert(&who, &registration.script, JobStatus::default());
+
+            if Self::match_job(&who, &registration)? {
+                // TODO improve event to contain list of matched sources
+                let job_id: JobId<T::AccountId> = (who.clone(), registration.script.clone());
+                Self::deposit_event(Event::JobRegistrationMatched(job_id));
+            }
+
+            Ok(().into())
+        }
+
+        /// Deregisters a job for the given script.
+        fn deregister_hook(who: &T::AccountId, script: &Script) -> Result<(), DispatchError> {
+            let job_status =
+                <StoredJobStatus<T>>::get(&who, &script).ok_or(Error::<T>::JobStatusNotFound)?;
+            ensure!(
+                job_status == JobStatus::Open,
+                Error::<T>::JobRegistrationUnmodifiable
+            );
+
+            <StoredJobStatus<T>>::remove(&who, &script);
+            Ok(().into())
+        }
+
+        /// Updates the allowed sources list of a [JobRegistration].
+        fn update_allowed_sources_hook(
+            who: &T::AccountId,
+            script: &Script,
+            _updates: &Vec<AllowedSourcesUpdate<T::AccountId>>,
+        ) -> Result<(), DispatchError> {
+            let job_status =
+                <StoredJobStatus<T>>::get(&who, &script).ok_or(Error::<T>::JobStatusNotFound)?;
+
+            ensure!(
+                job_status == JobStatus::Open,
+                Error::<T>::JobRegistrationUnmodifiable
+            );
+
+            Ok(().into())
+        }
+
+        /// Fulfills a previously registered job.
+        fn fulfill_hook(
+            who: &T::AccountId, // processor
+            fulfillment: &Fulfillment,
+            requester: <T::Lookup as StaticLookup>::Target, // the consumer that registered the job originally
+        ) -> Result<(), DispatchError> {
+            // find assignment
+            let job_id: JobId<T::AccountId> = (requester.clone(), fulfillment.script.clone());
+            <StoredJobAssignment<T>>::get(&job_id, &who)
+                .ok_or(Error::<T>::FulfillSourceNotAllowed)?;
+
+            // find job
+            let job_status = <StoredJobStatus<T>>::get(&who, &fulfillment.script)
+                .ok_or(Error::<T>::JobStatusNotFound)?;
+            let job_registration = <StoredJobRegistration<T>>::get(&job_id.0, &fulfillment.script)
+                .ok_or(Error::<T>::JobRegistrationNotFound)?;
+
+            // validate
+            ensure!(
+                job_status != JobStatus::Assigned,
+                Error::<T>::CannotFulfillJobWhenNotAssigned
+            );
+
+            T::RewardManager::pay_reward(
+                job_registration.reward.clone(),
+                T::Lookup::unlookup(who.clone()),
+            )
+            .map_err(|_| Error::<T>::FailedToPay)?;
```

Comment:
@gitsimon same here `T::RewardManager::pay_reward` is called already in the `fulfill` extrinsic in `pallet_acurast` before the `fulfill_hook` is called. So now it is called twice.

Reply:
RewardManager moved to marketplace